io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
-        if !readers.is_empty() {
-            self.nexus()
-                .children_iter()
-                .filter(|c| c.is_rebuilding())
-                .for_each(|c| match c.get_io_handle() {
-                    Ok(hdl) => {
-                        debug!(
-                            "{self:?}: connecting child device \
-                                in write-only mode: {c:?}"
-                        );
-                        writers.push(hdl);
+        match dev_ctx {
+            DrEvent::ChildRebuildBegin(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if !self.readers.is_empty() {
+                        // TODO: assert that the device is !is_writer here.
+                        match child.get_io_handle() {
+                            Ok(hdl) => {
+                                debug!("{self:?}: connecting child device in write-only mode: {child:?}");
+                                self.writers.push(hdl);
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
+                        // Make sure the child is not in readers. Most likely
+                        // the child won't be in readers
+                        // already here.
+                        self.readers
+                            .retain(|h| &h.get_device().device_name() != dev);
                     }
-                    Err(e) => {
-                        c.set_faulted_state(FaultReason::CantOpen);
-                        error!(
-                            "{self:?}: failed to get I/O handle \
-                                for {c:?}: {e}"
-                        );
+                }
+            }
+            DrEvent::ChildRebuildEnd(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if child.is_healthy() {
+                        // TODO: assert that the device is_writer and not
+                        // is_reader.
+                        match child.get_io_handle() {
+                            Ok(hdl) => {
+                                debug!("{self:?}: connecting child device as reader: {child:?}");
+                                self.readers.push(hdl);
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
                     }
-                });
+                }
+            }
```

Comment:
It seems that with this code now there's scope for duplicate entries in the lists, we should probably ensure that doesn't happen..

Reply:
The check would probably be based on iterating comparing the device name between existing NvmeDeviceHandle object and newly acquired one? vector length based check may not be completely reliable.

Reply:
I guess we could use the name from NvmeDeviceHandle?

Reply:
Added `is_reader() `and `is_writer()` helper functions in NexusChannel and used here.